File: 프로그래머스/3/468375. 발전소 회로 복구/발전소 회로 복구.java
import java.util.*;

class Solution {
    class State implements Comparable<State> {
        int mask, lastIdx, dist;
        State(int mask, int lastIdx, int dist) {
            this.mask = mask;
            this.lastIdx = lastIdx;
            this.dist = dist;
        }
        @Override
        public int compareTo(State o) {
            return this.dist - o.dist;
        }
    }

    int n, m, p;
    int[][] dists; 
    int[] elevatorDist;
    int[] prereqMask; //선행관계 마스크
    int INF = 100000000;

    public int solution(int h, String[] grid, int[][] panels, int[][] seqs) {
        this.n = grid.length; this.m = grid[0].length(); this.p = panels.length;

        int er = 0, ec = 0;
        Loop: for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i].charAt(j) == '@') {
                    er = i; ec = j;
                    break Loop;
                }
            }
        }

        dists = new int[p][p];
        elevatorDist = new int[p];
        for (int i = 0; i < p; i++) {
            int[][] dMap = bfsAll(grid, panels[i][1] - 1, panels[i][2] - 1);
            for (int j = 0; j < p; j++) {
                dists[i][j] = dMap[panels[j][1] - 1][panels[j][2] - 1];
            }
            elevatorDist[i] = dMap[er][ec];
        }

        prereqMask = new int[p];
        for (int[] seq : seqs) {
            prereqMask[seq[1] - 1] |= (1 << (seq[0] - 1));
        }

        return dijkstra(panels);
    }

    private int dijkstra(int[][] panels) {
        PriorityQueue<State> pq = new PriorityQueue<>();
        int[][] minDists = new int[1 << p][p];
        for (int[] row : minDists) Arrays.fill(row, INF);

        for (int i = 0; i < p; i++) {
            // 선행 조건이 없는 패널 i를 첫 번째로 활성화하러 감
            if (prereqMask[i] == 0) {
                int d;
                if (i == 0) {
                    // 1번 패널을 바로 켤 수 있다면 이동 거리 0
                    d = 0;
                } else {
                    // 1번 패널 위치(0번 인덱스)에서 i번 패널로 이동
                    d = getRealDist(0, i, panels);
                }
                
                if (minDists[1 << i][i] > d) {
                    minDists[1 << i][i] = d;
                    pq.offer(new State(1 << i, i, d));
                }
            }
        }

        int fullMask = (1 << p) - 1;
        while (!pq.isEmpty()) {
            State cur = pq.poll();

            if (cur.dist > minDists[cur.mask][cur.lastIdx]) continue;
            if (cur.mask == fullMask) return cur.dist;

            for (int next = 0; next < p; next++) {
                if ((cur.mask & (1 << next)) == 0 && (cur.mask & prereqMask[next]) == prereqMask[next]) {
                    int nextMask = cur.mask | (1 << next);
                    int d = getRealDist(cur.lastIdx, next, panels);
                    
                    if (minDists[nextMask][next] > cur.dist + d) {
                        minDists[nextMask][next] = cur.dist + d;
                        pq.offer(new State(nextMask, next, minDists[nextMask][next]));
                    }
                }
            }
        }
        return -1;
    }

    private int getRealDist(int u, int v, int[][] panels) {
        // u번째 패널 위치에서 v번째 패널 위치까지 이동 거리
        if (panels[u][0] == panels[v][0]) return dists[u][v];
        return elevatorDist[u] + Math.abs(panels[u][0] - panels[v][0]) + elevatorDist[v];
    }

    private int[][] bfsAll(String[] grid, int sr, int sc) {
        int[][] dist = new int[n][m];
        for (int[] row : dist) Arrays.fill(row, -1);
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{sr, sc});
        dist[sr][sc] = 0;
        int[] dr = {-1, 1, 0, 0}, dc = {0, 0, -1, 1};
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            for (int i = 0; i < 4; i++) {
                int nr = cur[0] + dr[i], nc = cur[1] + dc[i];
                if (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr].charAt(nc) != '#' && dist[nr][nc] == -1) {
                    dist[nr][nc] = dist[cur[0]][cur[1]] + 1;
                    q.offer(new int[]{nr, nc});
                }
            }
        }
        return dist;
    }
}

File: 프로그래머스/3/468375. 발전소 회로 복구/발전소 회로 복구.js
const map = new Map();

function canGo(next, seqs, visited) {
    for(let [a, b] of seqs) {
        if(next === b && visited[a] === 0) return false;
    }

    return true;
}

function bfs(ci, cj, ni, nj, grid) {
    const key = ci + "|" + cj + "|" + ni + "|" + nj;
    if(map.has(key)) return map.get(key);

    const arr = new Array(grid.length).fill(null).map(v => new Array(grid[0].length).fill(Infinity));
    const d = [[1, 0], [-1, 0], [0, 1], [0, -1]];

    const queue = [];
    queue.push([ci, cj]);
    arr[ci][cj] = 0;

    let head = 0;
    while(queue.length > head) {
        const [y, x] = queue[head++];
        for(let [dy, dx] of d) {
            const ny = y+dy, nx = x+dx;
            if(ny >= 0 && ny < grid.length && nx >= 0 && nx < grid[0].length) {
                if(grid[ny][nx] !== "#" && arr[ny][nx] === Infinity) {
                    queue.push([ny, nx]);
                    arr[ny][nx] = arr[y][x] + 1;
                }
            }
        }
    }

    map.set(key, arr[ni][nj]);
    return arr[ni][nj];
}

function getTime(current, next, grid, elevator) {
    const [cf, ci, cj] = current;
    const [nf, ni, nj] = next;
    const [ei, ej] = elevator;

    if(cf === nf) return bfs(ci, cj, ni, nj, grid);
    return bfs(ci, cj, ei, ej, grid) + Math.abs(cf-nf) + bfs(ei, ej, ni, nj, grid);
}

function getKey(arr) {
    let str = "";

    for(let v of arr) {
        str = str + v + "|";
    }

    return str;
}

function solution(h, grid, panels, seqs) {
    let answer = Infinity;
    let elevator = [];
    const map = new Map();
    const visited = new Array(panels.length).fill(0);
    panels = panels.map(v => [v[0], v[1]-1, v[2]-1]);
    seqs = seqs.map(v => [v[0]-1, v[1]-1]);

    for(let i = 0; i < grid.length; i++) {
        for(let j = 0; j < grid[0].length; j++) {
            if(grid[i][j] === "@") elevator = [i, j];
        }
    }

    function dfs(current, sum) {
        if(visited.every(v => v)) {
            answer = Math.min(answer, sum);
            return;
        }

        if(answer <= sum) return;

        const key1 = current;
        const key2 = getKey(visited);
        const key = key1 + key2;
        if(map.has(key) && map.get(key) <= sum) return;
        map.set(key, sum);

        for(let next = 0; next < panels.length; next++) {
            if(visited[next] || !canGo(next, seqs, visited)) continue;

            const time = getTime(panels[current], panels[next], grid, elevator);
            visited[next] = 1;
            dfs(next, sum+time);
            visited[next] = 0;
        }
    }

    dfs(0, 0);

    return answer;
}